### DISCUSSION TITLE
Communicate Back and Forth with Running AHK Script

### DISCUSSION THREAD

**Discussion by Lunatik6572** [spyoungtech/ahk#355] [Q&A] [answered]
Is there a way to start a script and communicate with the script at any given moment without ending the script?

I understand that you can use stdout in AHK to get the resulting string from `run_script()`, but this is when the script ends. I'm looking for a way to either invoke a python function (preferably with the option to call it with parameters) from the AHK script, or allow Python to get some variable from the currently running script without ending it.

The only solution I can think of is having AHK set the clipboard and have Python waiting for clipboard change but this seems very inefficient. Is there a better solution to this?

**Comment by spyoungtech** [comment, marked as answer]
Sorry for the late response on this. Basically, this is what _this_ library does. It invokes an AHK script as a subprocess and communicates back and forth with it via stdin/stdout. You can poke around the code in this repo and see how it's implemented.

When you use this library to add Python functions as hotkey callbacks -- When the AHK hotkey is invoked, it sends messages to Python via stdout, which Python in turn reads and the contents of the message indicates which Python callback function should be called. When you call a method like `win_get` Python [sends a message](https://github.com/spyoungtech/ahk/blob/70f43c4b752fa2fc3ff6ea984759229b8c26470f/ahk/_sync/transport.py#L722-L723) containing the function parameters to the AHK subprocess via stdin; the AHK script reads stdin to know which function to call and then returns a message via stdout that Python [reads and translates into a return value](https://github.com/spyoungtech/ahk/blob/70f43c4b752fa2fc3ff6ea984759229b8c26470f/ahk/_sync/transport.py#L724-L745). 



One limitation of AHK here, however, is that it lacks threading capabilities and blocking actions (like awaiting input from stdin) completely prevent any background messages from being processed in AHK (which are used for things like timers, hotkeys, and more). Because of this, simultaneous bidirectional communication using stdin/stoud is not practical within a single AHK process. So, an implementation detail in this library is that this is worked around with two processes: one for receiving commands from Python (like for method calls like `win_get`, `mouse_move`, etc.) and another process for sending commands to Python (like in response to Hotkeys, `OnClipboardChange`, etc.). This is discussed at some length in #274 - including alternative methods of communication that may avoid this.

[These ~30 lines of AHK code](https://github.com/spyoungtech/ahk/blob/70f43c4b752fa2fc3ff6ea984759229b8c26470f/ahk/templates/daemon-v2.ahk#L2984-L3020) (ignoring the jinja blocks) is essentially what lets AHK receive inputs from Python and return responses. You could reasonably invert this relationship to have AHK call Python functions and receive its results, too. For hotkeys and clipboard callbacks, [these lines](https://github.com/spyoungtech/ahk/blob/70f43c4b752fa2fc3ff6ea984759229b8c26470f/ahk/_hotkey.py#L340-L351) and [these lines](https://github.com/spyoungtech/ahk/blob/70f43c4b752fa2fc3ff6ea984759229b8c26470f/ahk/_hotkey.py#L272-L308) of Python are what are responsible for receiving messages [sent from AHK](https://github.com/spyoungtech/ahk/blob/70f43c4b752fa2fc3ff6ea984759229b8c26470f/ahk/templates/hotkeys-v2.ahk#L70-L75). There are a number of other details involved that make this protocol work (like message formats, and so on), but those are the basics.

Lastly, this library also provides means for [writing extensions](https://ahk.readthedocs.io/en/latest/extending.html) to extend the functionality by adding your own AHK code to the daemon that accepts messages from Python. So you can write a bit of AHK code and leverage the communication framework already implemented by this library. Depending on the exact details, this might be a way that you can implement the back-and-forth communication you describe for your use-case. See one [simple example](https://gist.github.com/spyoungtech/185f8cb0603147f6e84fcc012b2cd96c) of an extension that manipulates state in AHK and later can retrieve that state from Python. This is a really simple example, but may put you in the right direction.


I hope that helps answer your question.

**Reply by Lunatik6572** [reply]
I appreciate the detailed response! The way you communicate with the AHK is a very creative solution. After looking at all the examples you gave, it seems like writing an extension may actually best suit my needs. 

My project uses AHK UIA to get values from different windows. Python then gets those values and does it's own thing afterwards. The issue is that I pass variables to the AHK script by using string replacement after loading the script as a string in python. I then use `run_script()` and wait for a change in the clipboard. This works 99% of the time as expected. But there are a few times where AHK seems to lag for several seconds (usually no longer than 10s) before it actually runs the code, some times it fails, and other times it fails to set the clipboard.

If there was a way to have the script always running and loaded and then just invoke the necessary functions and return the data, then I believe this would solve my Issue. I am not sure if the extensions are loaded once and then just waiting to be invoked, but the example you gave and the example I built seems to confirm that it is. This sounds very promising.

I'm going to try and continue with extensions and see if this solves all the other issues! Thank you!
